File: Tests/FocusCoreTests/FocusScriptBuilderTests.swift
import FocusCore
#if canImport(JavaScriptCore)
import JavaScriptCore
#endif
import Testing

struct FocusScriptBuilderTests {
    @Test
    func documentStartScriptIncludesConfigAndPhase() throws {
        let script = try FocusScriptBuilder.makeUserScript(
            for: .documentStart,
            settings: FocusSettings(debugMode: true)
        )

        #expect(script.contains("\"debugMode\":true"))
        #expect(script.contains("documentStart"))
        #expect(script.contains("width=device-width"))
    }

    @Test
    func dynamicRuleMatchesExpectedPage() {
        let dynamicRule = FocusRuleCatalog.defaultRules.first { $0.id == "dynamic-prune" }

        #expect(dynamicRule?.match(host: "t.bilibili.com", path: "/") == true)
        #expect(dynamicRule?.match(host: "www.bilibili.com", path: "/") == false)
    }

    #if canImport(JavaScriptCore)
    @Test
    func featureScriptsCompileAsJavaScript() throws {
        let context = try #require(JSContext())

        for rule in FocusRuleCatalog.defaultRules {
            for feature in rule.features {
                guard let script = feature.script else {
                    continue
                }

                context.exception = nil
                let wrappedScript = "(function(config, helpers) {\n\(script)\n})"
                let compiledFunction = context.evaluateScript(wrappedScript)

                if compiledFunction == nil || context.exception != nil {
                    let exceptionMessage = context.exception?.toString() ?? "unknown JavaScript error"
                    Issue.record("Failed to compile \(rule.id)/\(feature.featureId): \(exceptionMessage)")
                }

                #expect(compiledFunction != nil)
                #expect(context.exception == nil)
            }
        }
    }
    #endif
}

File: focus-core/src/commonTest/kotlin/org/bilibilifocus/core/injection/FocusScriptBuilderTests.kt
package org.bilibilifocus.core.injection

import org.bilibilifocus.core.model.FocusPageRule
import org.bilibilifocus.core.model.FocusSettings
import kotlin.test.Test
import kotlin.test.assertTrue

class FocusScriptBuilderTests {

    @Test
    fun `document start script includes config and phase`() {
        val script = FocusScriptBuilder.makeUserScript(
            phase = FocusPageRule.RunPhase.DOCUMENT_START,
            settings = FocusSettings(debugMode = true),
        )

        // Script should contain the debug flag
        assertTrue("\"debugMode\":true" in script, "Script should contain debugMode setting")
        // Script should contain the phase marker
        assertTrue("documentStart" in script, "Script should contain documentStart phase")
        // Script should contain viewport meta
        assertTrue("width=device-width" in script, "Script should contain viewport meta")
    }

    @Test
    fun `dynamic rule matches expected page`() {
        val dynamicRule = FocusRuleCatalog.defaultRules.firstOrNull { it.id == "dynamic-prune" }

        assertTrue(dynamicRule != null, "Should find dynamic-prune rule")
        assertTrue(dynamicRule!!.match(host = "t.bilibili.com", path = "/"))
        assertTrue(!dynamicRule.match(host = "www.bilibili.com", path = "/"))
    }

    @Test
    fun `global viewport rule matches multiple hosts`() {
        val viewportRule = FocusRuleCatalog.defaultRules.firstOrNull { it.id == "global-viewport" }

        assertTrue(viewportRule != null)
        assertTrue(viewportRule!!.match(host = "www.bilibili.com", path = "/video/BV123"))
        assertTrue(viewportRule.match(host = "t.bilibili.com", path = "/"))
        assertTrue(viewportRule.match(host = "search.bilibili.com", path = "/all"))
    }
}
